importer/src/main/java/org/hiero/mirror/importer/downloader/block/BlockNode.java

```diff
@@ -0,0 +1,250 @@
+
+        try {
+            var assembler = new BlockAssembler();
+            var request = SubscribeStreamRequest.newBuilder()
+                    .setEndBlockNumber(INFINITE_END_BLOCK_NUMBER)
+                    .setStartBlockNumber(blockNumber)
+                    .build();
+            grpcCall.set(ClientCalls.blockingV2ServerStreamingCall(
+                    channel,
+                    BlockStreamSubscribeServiceGrpc.getSubscribeBlockStreamMethod(),
+                    CallOptions.DEFAULT,
+                    request));
+            var timeout = new TimeoutContext(blockTimeout);
+            SubscribeStreamResponse response;
+
+            outer:
+            while ((response = grpcCall.get().read(timeout.remaining(), TimeUnit.MILLISECONDS)) != null) {
+                switch (response.getResponseCase()) {
+                    case BLOCK_ITEMS -> {
+                        var streamedBlock = assembler.assemble(response.getBlockItems());
+                        if (streamedBlock != null) {
+                            timeout.reset();
+                            onStreamedBlock.accept(streamedBlock);
+                        }
+                    }
+                    case STATUS -> {
+                        var status = response.getStatus();
+                        if (status == SubscribeStreamResponse.Code.READ_STREAM_SUCCESS) {
+                            // The server may end the stream gracefully for various reasons, and this shouldn't be
+                            // treated as an error.
+                            log.info("Block server ended the subscription with {}", status);
+                            break outer;
+                        }
+
+                        throw new BlockStreamException("Received status " + response.getStatus() + " from block node");
+                    }
+                    default -> throw new BlockStreamException("Unknown response case " + response.getResponseCase());
+                }
+            }
+        } catch (Exception ex) {
+            onError();
+
+            if (ex instanceof BlockStreamException blockStreamException) {
+                throw blockStreamException;
+            }
+
+            throw new BlockStreamException(ex);
+        } finally {
+            if (grpcCall.get() != null) {
+                grpcCall.get().cancel("unsubscribe", null);
+            }
+        }
+    }
+
+    @Override
+    public String toString() {
+        return String.format("BlockNode(%s)", properties.getEndpoint());
+    }
+
+    public BlockNode tryReadmit(boolean force) {
+        if (!active && (force || Instant.now().isAfter(readmitTime.get()))) {
+            active = true;
+        }
+
+        return this;
+    }
+
+    private void onError() {
+        if (++errors >= streamProperties.getMaxSubscribeAttempts()) {
+            log.warn(
+                    "Failed to stream blocks from {} {} times consecutively, mark it inactive",
```

Comment:
Add error message to end and reword.

```suggestion
                    "Marking connection to {} as inactive after {} attempts: {}"
```